### PR TITLE
Add company view

### PR DESCRIPTION
**Pull request by alfredoct96**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL |  /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | /
| Deprecations? | /

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Added view page for companies.
It includes:
* Basic company's info
* Button to export company's info
* Graph with sum of engagement and points of each company's contact
* Table with pagination of all contacts of company

The request so includes all routes to get all contacts of company and export company info.

I added also methods for 
* view company page
* get sum of engagements and score of each company's user
* export company's info

mautic1

mautic2

#### Steps to test this PR:
1. Load up [this PR](https://m3.mautibox.com)
2. Create a company and assign some users to it
3. Go to Company menù and click to a company, it will take you to company view page.

### CODE REVIEW

**Review by ShareBtech — COMMENTED** [review]
Began by adding a new custom field: Company Category

When testing adding details to the company got the following error:
 500 Internal Server Error - An exception occurred while executing 'UPDATE 8936_companies SET companyaddress1 = ?, companyaddress2 = ?, companyemail = ?, companyphone = ?, companyzipcode = ?, companywebsite = ?, company_category = ?, companyfax = ?, companyindustry = ? WHERE id = ?' with params ["This is address 1", "This is address 2", "info@amazon.world", "+180099990000", "ZIP09822", "htttps:\/\/www.amazong.com", "Market Place", "11", null, 3]: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'companyfax' in 'field list' 


Doctrine\DBAL\Exception\InvalidFieldNameException



    /vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:60 at
    /vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:166 at Doctrine\DBAL\Driver\AbstractMySQLDriver -> convertException ( 'An exception occurred while executing 'UPDATE 8936_companies SET companyaddress1 = ?, companyaddress2 = ?, companyemail = ?, companyphone = ?, companyzipcode = ?, companywebsite = ?, company_category = ?, companyfax = ?, companyindustry = ? WHERE id = ?' with params ["This is address 1", "This is address 2", "info@amazon.world", "+180099990000", "ZIP09822", "htttps:\/\/www.amazong.com", "Market Place", "11", null, 3]: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'companyfax' in 'field list'', object(PDOException) )
    /vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:146 at Doctrine\DBAL\DBALException :: wrapException ( object(Driver), object(PDOException), 'An exception occurred while executing 'UPDATE 8936_companies SET companyaddress1 = ?, companyaddress2 = ?, companyemail = ?, companyphone = ?, companyzipcode = ?, companywebsite = ?, company_category = ?, companyfax = ?, companyindustry = ? WHERE id = ?' with params ["This is address 1", "This is address 2", "info@amazon.world", "+180099990000", "ZIP09822", "htttps:\/\/www.amazong.com", "Market Place", "11", null, 3]: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'companyfax' in 'field list'' )
    /vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:1069 at Doctrine\DBAL\DBALException :: driverExceptionDuringQuery ( object(Driver), object(PDOException), 'UPDATE 8936_companies SET companyaddress1 = ?, companyaddress2 = ?, companyemail = ?, companyphone = ?, companyzipcode = ?, companywebsite = ?, company_category = ?, companyfax = ?, companyindustry = ? WHERE id = ?', array('This is address 1', 'This is address 2', 'info@amazon.world', '+180099990000', 'ZIP09822', 'htttps://www.amazong.com', 'Market Place', '11', null, '3') )
    /vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:743 at Doctrine\DBAL\Connection -> executeUpdate ( 'UPDATE 8936_companies SET companyaddress1 = ?, companyaddress2 = ?, companyemail = ?, companyphone = ?, companyzipcode = ?, companywebsite = ?, company_category = ?, companyfax = ?, companyindustry = ? WHERE id = ?', array('This is address 1', 'This is address 2', 'info@amazon.world', '+180099990000', 'ZIP09822', 'htttps://www.amazong.com', 'Market Place', '11', null, '3'), array() )
    /app/bundles/LeadBundle/Entity/CustomFieldRepositoryTrait.php:270 at Doctrine\DBAL\Connection -> update ( '8936_companies', array('companyaddress1' => 'This is address 1', 'companyaddress2' => 'This is address 2', 'companyemail' => 'info@amazon.world', 'companyphone' => '+180099990000', 'companyzipcode' => 'ZIP09822', 'companywebsite' => 'htttps://www.amazong.com', 'company_category' => 'Market Place', 'companyfax' => '11', 'companyindustry' => null), array('id' => '3') )
    /app/bundles/CoreBundle/Model/FormModel.php:111 at Mautic\LeadBundle\Entity\CompanyRepository -> saveEntity ( object(Company) )
    /app/bundles/LeadBundle/Model/CompanyModel.php:87 at Mautic\CoreBundle\Model\FormModel -> saveEntity ( object(Company), false )
    /app/bundles/LeadBundle/Controller/CompanyController.php:403 at Mautic\LeadBundle\Model\CompanyModel -> saveEntity ( object(Company), false )
    /app/bundles/CoreBundle/Controller/CommonController.php:449 at Mautic\LeadBundle\Controller\CompanyController -> editAction ( '3', '' )
    /vendor/symfony/http-kernel/HttpKernel.php:151 at Mautic\CoreBundle\Controller\CommonController -> executeAction ( 'edit', '3', '0', '' )
    /vendor/symfony/http-kernel/HttpKernel.php:68 at Symfony\Component\HttpKernel\HttpKernel -> handleRaw ( object(Request), '1' )
    /vendor/symfony/http-kernel/Kernel.php:200 at Symfony\Component\HttpKernel\HttpKernel -> handle ( object(Request), '1', true )
    /app/AppKernel.php:104 at Symfony\Component\HttpKernel\Kernel -> handle ( object(Request), '1', true )
    /app/middlewares/CORSMiddleware.php:91 at AppKernel -> handle ( object(Request), '1', true )
    /app/middlewares/CatchExceptionMiddleware.php:43 at Mautic\Middleware\CORSMiddleware -> handle ( object(Request), '1', true )
    /app/middlewares/Dev/IpRestrictMiddleware.php:64 at Mautic\Middleware\CatchExceptionMiddleware -> handle ( object(Request), '1', true )
    /app/middlewares/VersionCheckMiddleware.php:67 at Mautic\Middleware\Dev\IpRestrictMiddleware -> handle ( object(Request), '1', true )
    /app/middlewares/TrustMiddleware.php:51 at Mautic\Middleware\VersionCheckMiddleware -> handle ( object(Request), '1', true )
    /vendor/stack/builder/src/Stack/StackedHttpKernel.php:23 at Mautic\Middleware\TrustMiddleware -> handle ( object(Request), '1', true )
    /vendor/stack/run/src/Stack/run.php:13 at Stack\StackedHttpKernel -> handle ( object(Request) )
    /index_dev.php:25 at Stack\run ( object(StackedHttpKernel) )
    /index.php:6 at require_once ( '/var/app/current/code/pulls/8936/index_dev.php' )

**Comment by RCheesley** [comment]
@ShareBtech I'm pretty sure this is an issue with the sample data due to a duplicated field, pinging @heathdutton who runs Mautibox to double check!

**Review by RCheesley — DISMISSED** [review]
Thanks for this great PR @alfredoct96 - it is really helpful to be able to get a company overview!

One thing that I think others would find useful is being able to manipulate the user list - for example be able to select multiple users and update them as you can from the contact list itself. Is that something you feel able to add?  If not perhaps that could be a later enhancement as the PR is a great addition!

If we merge this in 3.1 the next opportunity for enhancements would be three months later, so it would be nice to get that included in this PR if at all possible!

**Comment by alfredoct96** [comment]
> Began by adding a new custom field: Company Category
> 
> When testing adding details to the company got the following error:
> 500 Internal Server Error - An exception occurred while executing 'UPDATE 8936_companies SET companyaddress1 = ?, companyaddress2 = ?, companyemail = ?, companyphone = ?, companyzipcode = ?, companywebsite = ?, company_category = ?, companyfax = ?, companyindustry = ? WHERE id = ?' with params ["This is address 1", "This is address 2", "[info@amazon.world](mailto:info@amazon.world)", "+180099990000", "ZIP09822", "htttps://www.amazong.com", "Market Place", "11", null, 3]: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'companyfax' in 'field list'
> 
> Doctrine\DBAL\Exception\InvalidFieldNameException
> 
> ```
> /vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:60 at
> /vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:166 at Doctrine\DBAL\Driver\AbstractMySQLDriver -> convertException ( 'An exception occurred while executing 'UPDATE 8936_companies SET companyaddress1 = ?, companyaddress2 = ?, companyemail = ?, companyphone = ?, companyzipcode = ?, companywebsite = ?, company_category = ?, companyfax = ?, companyindustry = ? WHERE id = ?' with params ["This is address 1", "This is address 2", "info@amazon.world", "+180099990000", "ZIP09822", "htttps:\/\/www.amazong.com", "Market Place", "11", null, 3]: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'companyfax' in 'field list'', object(PDOException) )
> /vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:146 at Doctrine\DBAL\DBALException :: wrapException ( object(Driver), object(PDOException), 'An exception occurred while executing 'UPDATE 8936_companies SET companyaddress1 = ?, companyaddress2 = ?, companyemail = ?, companyphone = ?, companyzipcode = ?, companywebsite = ?, company_category = ?, companyfax = ?, companyindustry = ? WHERE id = ?' with params ["This is address 1", "This is address 2", "info@amazon.world", "+180099990000", "ZIP09822", "htttps:\/\/www.amazong.com", "Market Place", "11", null, 3]: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'companyfax' in 'field list'' )
> /vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:1069 at Doctrine\DBAL\DBALException :: driverExceptionDuringQuery ( object(Driver), object(PDOException), 'UPDATE 8936_companies SET companyaddress1 = ?, companyaddress2 = ?, companyemail = ?, companyphone = ?, companyzipcode = ?, companywebsite = ?, company_category = ?, companyfax = ?, companyindustry = ? WHERE id = ?', array('This is address 1', 'This is address 2', 'info@amazon.world', '+180099990000', 'ZIP09822', 'htttps://www.amazong.com', 'Market Place', '11', null, '3') )
> /vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:743 at Doctrine\DBAL\Connection -> executeUpdate ( 'UPDATE 8936_companies SET companyaddress1 = ?, companyaddress2 = ?, companyemail = ?, companyphone = ?, companyzipcode = ?, companywebsite = ?, company_category = ?, companyfax = ?, companyindustry = ? WHERE id = ?', array('This is address 1', 'This is address 2', 'info@amazon.world', '+180099990000', 'ZIP09822', 'htttps://www.amazong.com', 'Market Place', '11', null, '3'), array() )
> /app/bundles/LeadBundle/Entity/CustomFieldRepositoryTrait.php:270 at Doctrine\DBAL\Connection -> update ( '8936_companies', array('companyaddress1' => 'This is address 1', 'companyaddress2' => 'This is address 2', 'companyemail' => 'info@amazon.world', 'companyphone' => '+180099990000', 'companyzipcode' => 'ZIP09822', 'companywebsite' => 'htttps://www.amazong.com', 'company_category' => 'Market Place', 'companyfax' => '11', 'companyindustry' => null), array('id' => '3') )
> /app/bundles/CoreBundle/Model/FormModel.php:111 at Mautic\LeadBundle\Entity\CompanyRepository -> saveEntity ( object(Company) )
> /app/bundles/LeadBundle/Model/CompanyModel.php:87 at Mautic\CoreBundle\Model\FormModel -> saveEntity ( object(Company), false )
> /app/bundles/LeadBundle/Controller/CompanyController.php:403 at Mautic\LeadBundle\Model\CompanyModel -> saveEntity ( object(Company), false )
> /app/bundles/CoreBundle/Controller/CommonController.php:449 at Mautic\LeadBundle\Controller\CompanyController -> editAction ( '3', '' )
> /vendor/symfony/http-kernel/HttpKernel.php:151 at Mautic\CoreBundle\Controller\CommonController -> executeAction ( 'edit', '3', '0', '' )
> /vendor/symfony/http-kernel/HttpKernel.php:68 at Symfony\Component\HttpKernel\HttpKernel -> handleRaw ( object(Request), '1' )
> /vendor/symfony/http-kernel/Kernel.php:200 at Symfony\Component\HttpKernel\HttpKernel -> handle ( object(Request), '1', true )
> /app/AppKernel.php:104 at Symfony\Component\HttpKernel\Kernel -> handle ( object(Request), '1', true )
> /app/middlewares/CORSMiddleware.php:91 at AppKernel -> handle ( object(Request), '1', true )
> /app/middlewares/CatchExceptionMiddleware.php:43 at Mautic\Middleware\CORSMiddleware -> handle ( object(Request), '1', true )
> /app/middlewares/Dev/IpRestrictMiddleware.php:64 at Mautic\Middleware\CatchExceptionMiddleware -> handle ( object(Request), '1', true )
> /app/middlewares/VersionCheckMiddleware.php:67 at Mautic\Middleware\Dev\IpRestrictMiddleware -> handle ( object(Request), '1', true )
> /app/middlewares/TrustMiddleware.php:51 at Mautic\Middleware\VersionCheckMiddleware -> handle ( object(Request), '1', true )
> /vendor/stack/builder/src/Stack/StackedHttpKernel.php:23 at Mautic\Middleware\TrustMiddleware -> handle ( object(Request), '1', true )
> /vendor/stack/run/src/Stack/run.php:13 at Stack\StackedHttpKernel -> handle ( object(Request) )
> /index_dev.php:25 at Stack\run ( object(StackedHttpKernel) )
> /index.php:6 at require_once ( '/var/app/current/code/pulls/8936/index_dev.php' )
> ```

@ShareBtech  @RCheesley  Seems that database is not aligned with branch.
When try to update a company the important error is "Column not found: 1054 Unknown column 'companyfax' in 'field list'", seems that column companyfax doesn't exists, i don't think is a problem of my pull request .

**Comment by alfredoct96** [comment]
> Thanks for this great PR @alfredoct96 - it is really helpful to be able to get a company overview!
> 
> One thing that I think others would find useful is being able to manipulate the user list - for example be able to select multiple users and update them as you can from the contact list itself. Is that something you feel able to add? If not perhaps that could be a later enhancement as the PR is a great addition!
> 
> If we merge this in 3.1 the next opportunity for enhancements would be three months later, so it would be nice to get that included in this PR if at all possible!

Thanks @RCheesley !  Ill'try to add it in this PR!

**Comment by kuzmany** [comment]
Audit logs tab would be great addition for this PR

**Comment by alfredoct96** [comment]
> Thanks for this great PR @alfredoct96 - it is really helpful to be able to get a company overview!
> 
> One thing that I think others would find useful is being able to manipulate the user list - for example be able to select multiple users and update them as you can from the contact list itself. Is that something you feel able to add? If not perhaps that could be a later enhancement as the PR is a great addition!
> 
> If we merge this in 3.1 the next opportunity for enhancements would be three months later, so it would be nice to get that included in this PR if at all possible!



> Thanks for this great PR @alfredoct96 - it is really helpful to be able to get a company overview!
> 
> One thing that I think others would find useful is being able to manipulate the user list - for example be able to select multiple users and update them as you can from the contact list itself. Is that something you feel able to add? If not perhaps that could be a later enhancement as the PR is a great addition!
> 
> If we merge this in 3.1 the next opportunity for enhancements would be three months later, so it would be nice to get that included in this PR if at all possible!

@RCheesley  I've added the changes requested! You can try it

**Comment by RCheesley** [comment]
@alfredoct96 could you rebase against staging? Looks like it is a bit out of sync.

**Comment by alfredoct96** [comment]
> @alfredoct96 could you rebase against staging? Looks like it is a bit out of sync.

Hi @RCheesley, i synchronized it with staging, have a good day.

**Comment by npracht** [comment]
@kuzmany and @RCheesley i like your suggestions with more features, but i don't think we should postpone this merge because we would like more feature. This is already a nice improvement and we can talk extra improvement on forums for future versions.

@alfredoct96 you absolutely need to add unit tests to your code to make it mergeable. You can get inspired from what have been done on the contact overview page. Can you do that ?

**Comment by npracht** [comment]
I'll carry the documentation :)

**Comment by RCheesley** [comment]
> I'll carry the documentation :) 🙇‍♀️

**Comment by RCheesley** [comment]
> @kuzmany and @RCheesley i like your suggestions with more features, but i don't think we should postpone this merge because we would like more feature. This is already a nice improvement and we can talk extra improvement on forums for future versions.

Absolutely, we can create an issue to add those features in a subsequent feature release, 3.2 maybe. But something is better than nothing! :)

**Comment by RCheesley** [comment]
> @alfredoct96 you absolutely need to add unit tests to your code to make it mergeable. You can get inspired from what have been done on the contact overview page. Can you do that ?

If you need any help or guidance with getting the test coverage up please don't hesitate to give us a shout - we can help you with this. Would love to see this merged in 3.1!

**Review by npracht — DISMISSED** [review]
Just tested, and it seems robust, this is an awesome improvement to do some ABM !!! 👍

File: app/bundles/LeadBundle/Views/Company/company.html.php [hunk 1]
```diff
@@ -0,0 +1,293 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (kuzmany) [inline app/bundles/LeadBundle/Views/Company/company.html.php]: All new files should be marked with actual year for copyright (2020)

**Review by kuzmany — DISMISSED** [review]
Just change the years of new files.
Looks good. Tested and works properly. 
I like this PR so much. Good job!

**Comment by npracht** [comment]
Documentation is ready --> https://github.com/mautic/mautic-documentation/pull/105

File: app/bundles/LeadBundle/Views/Company/company.html.php (hunk 1 above)
Comment (alfredoct96) [reply app/bundles/LeadBundle/Views/Company/company.html.php]: I changed the year!

**Comment by alfredoct96** [comment]
Thanks @npracht  @RCheesley ! I added the tests and codecoverage passed!

Tell me if something else is missing or is all ok!

**Comment by npracht** [comment]
@dennisameling id tests are okay, it would be an awesome feature for a minor version :)

**Comment by RCheesley** [comment]
@alfredoct96 great, the code coverage report seems a bit wonky, possibly because PR's were merged while the Travis build was executing (we had a bit of a backlog due to the sprint!) - could you rebase once more and we'll hold fire on any more merges until the build completes.

**Comment by alfredoct96** [comment]
> @alfredoct96 great, the code coverage report seems a bit wonky, possibly because PR's were merged while the Travis build was executing (we had a bit of a backlog due to the sprint!) - could you rebase once more and we'll hold fire on any more merges until the build completes.



> @alfredoct96 great, the code coverage report seems a bit wonky, possibly because PR's were merged while the Travis build was executing (we had a bit of a backlog due to the sprint!) - could you rebase once more and we'll h

**Comment by alfredoct96** [comment]
> @alfredoct96 great, the code coverage report seems a bit wonky, possibly because PR's were merged while the Travis build was executing (we had a bit of a backlog due to the sprint!) - could you rebase once more and we'll hold fire on any more merges until the build completes.

@RCheesley  No problem, but since the pull request is large enough, according to the documentation, in order to not make mistakes I considered the merge with staging and not the rebase

**Review by dennisameling — APPROVED** [review]
Code LGTM 👍 thanks @alfredoct96, awesome contribution!!

**Comment by alfredoct96** [comment]
> Code LGTM  thanks @alfredoct96, awesome contribution!!

It's a pleasure!

**Pull request merged**